### PR TITLE
SDK should update certificate pools safely

### LINKED ISSUE #23: Update safely a certificate pool with GetConfigForClient
The SDK does not remove the old roots as it's not possible to safely change the RootCAs/ClientCAs directly on the tls.Config. But tls.Config method GetConfigForClient allows to get a custom tls.Config after a ClientHello. More information:
https://tip.golang.org/pkg/crypto/tls/#Config.GetConfigForClient
https://diogomonica.com/2017/01/11/hitless-tls-certificate-rotation-in-go/

### PR DESCRIPTION
**Pull request by maraino**
### Description
Certificate pools should be safely updated when root or federation certificates are rotated. A server uses `tls.Config.GetConfigForClient` to provide the refreshed version of the tls.Config while clients use a custom http.Transport.DialTLS that uses the refreshed tls.Config. Fixes #23

This PR also adds the method `BootstrapListener` that creates a `net.Listener` configured with the TLS certificate provided by the CA

### CODE REVIEW

File: ca/bootstrap_test.go [hunk 1]
```diff
@@ -530,3 +530,71 @@ func doReload(ca *CA) error {
...
+			}()
+			defer wg.Wait()
+			defer lis.Close()
+
+			client, err := BootstrapClient(context.Background(), token())
+			if err != nil {
+				t.Errorf("BootstrapClient() error = %v", err)
+				return
+			}
+			println("https://" + lis.Addr().String())
```
Comment (dopey) [inline ca/bootstrap_test.go]: unnecessary println?

File: ca/mutable_tls_config.go [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+}
+
+// Init initializes the mutable tls.Config with the given tls.Config.
+func (c *mutableTLSConfig) Init(base *tls.Config) {
+	c.Lock()
+	c.config = base.Clone()
+	c.Unlock()
+}
+
+// TLSConfig returns the updated tls.Config it it has changed. It's is used in
```
Comment (dopey) [inline ca/mutable_tls_config.go]: It's used

File: ca/mutable_tls_config.go [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+	// Set new pool
+	c.Lock()
+	c.config.RootCAs = rootCAs
+	c.config.ClientCAs = clientCAs
+	c.mutRootCerts = []*x509.Certificate{}
+	c.mutClientCerts = []*x509.Certificate{}
+	c.Unlock()
+}
+
+// AddFixedClientCACert add an in-mutable cert to ClientCAs.
```
Comment (dopey) [inline ca/mutable_tls_config.go]: I think you're going for Immutable (not Inmutable or Fixed)

File: ca/mutable_tls_config.go [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+	// Set new pool
+	c.Lock()
+	c.config.RootCAs = rootCAs
+	c.config.ClientCAs = clientCAs
+	c.mutRootCerts = []*x509.Certificate{}
+	c.mutClientCerts = []*x509.Certificate{}
+	c.Unlock()
+}
+
+// AddImmutableClientCACert add an in-mutable cert to ClientCAs.
```
Comment (dopey) [inline ca/mutable_tls_config.go]: in-mutable -> immutable

File: ca/mutable_tls_config.go [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+func (c *mutableTLSConfig) AddImmutableClientCACert(cert *x509.Certificate) {
+	c.Lock()
+	c.clientCerts = append(c.clientCerts, cert)
+	c.Unlock()
+}
+
+// AddImmutableRootCACert add an in-mutable cert to RootCas.
```
Comment (dopey) [inline ca/mutable_tls_config.go]: in-mutable -> immutable

File: ca/tls.go [hunk 6]
```diff
@@ -139,7 +169,7 @@ func IntermediateCertificate(sign *api.SignResponse) (*x509.Certificate, error)
 
 // RootCertificate returns the root certificate from the sign response.
 func RootCertificate(sign *api.SignResponse) (*x509.Certificate, error) {
-	if sign.TLS == nil || len(sign.TLS.VerifiedChains) == 0 {
+	if sign == nil || sign.TLS == nil || len(sign.TLS.VerifiedChains) == 0 {
 		return nil, errors.New("ca: certificate does not exists")
```
Comment (dopey) [inline ca/tls.go]: certificate does not exist$ (not exists)

File: ca/mutable_tls_config.go (hunk 4 above)
Comment (maraino) [reply ca/mutable_tls_config.go]: I fixed this yesterday

File: ca/mutable_tls_config.go (hunk 5 above)
Comment (maraino) [reply ca/mutable_tls_config.go]: I fixed this yesterday.

File: ca/tls.go (hunk 6 above)
Comment (maraino) [reply ca/tls.go]: Fixed with  [f1f6c54](/smallstep/certificates/pull/27/commits/f1f6c548ad2eaf7338393cf964938adb8d4b0295)

**Pull request merged**